### PR TITLE
Allow Email rules to have much more props and autocomplete in editor

### PR DESCRIPTION
**Pull request by Kukks**
https://i.imgur.com/zQtGKBC.gif

### CODE REVIEW

**Comment by dennisreimann** [comment]
Is this still a kukksification or some kind of voodoo already? :D

**Comment by pavlenex** [comment]
Interesting, as Dennis said, let us know where this one is heading and when reviewers/helps are needed.

**Comment by Kukks** [comment]
> Interesting, as Dennis said, let us know where this one is heading and when reviewers/helps are needed.

I basically took one of the core mechanisms of BTC Transmuter and plugged it in the email rules to make it extra flexible. It will take the invoice model, currently the Invoice model from Greenfield API, and allow the user to use any field of it, along with some additional "script" magic. For example you can do {Invoice.Amount +5} or {Invoice.Currency.ToUpper()}

File: BTCPayServer/HostedServices/StoreEmailRuleProcessorSender.cs [hunk 1]
```diff
@@ -79,25 +82,27 @@ protected override async Task ProcessEvent(object evt, CancellationToken cancell
...
-                decimal.Round(i.Amount, _currencyNameTable.GetCurrencyData(i.Currency, true).Divisibility,
-                    MidpointRounding.ToEven).ToString(CultureInfo.InvariantCulture))
-            .Replace("{Invoice.Currency}", i.Currency)
-            .Replace("{Invoice.Status}", i.Status.ToString())
-            .Replace("{Invoice.AdditionalStatus}", i.AdditionalStatus.ToString())
-            .Replace("{Invoice.OrderId}", i.Metadata.ToObject<InvoiceMetadata>().OrderId);
+        return Regex.Replace(value, @"{(?<exp>[^}]+)}", match => {
+            var p = Expression.Parameter(typeof(T), propName);
+            var e = System.Linq.Dynamic.Core.DynamicExpressionParser.ParseLambda(new[] { p }, null, match.Groups["exp"].Value);
+            return (e.Compile().DynamicInvoke(item) ?? value).ToString();
```
Comment (NicolasDorier) [inline BTCPayServer/HostedServices/StoreEmailRuleProcessorSender.cs:100]: nope nope nope nope, man there is code injection here.

**Comment by NicolasDorier** [comment]
I really like it, but the magic with the dynamic LINQ need to go.

I believe this PR open a code injection vulnerability, where anybody can inject arbitrary C# to be executed server side.

**Comment by Kukks** [comment]
we now support jsonath on the metadata without this crazy reflection

**Pull request closed without merging**